lib/puppet/functions/deprecation.rb

```diff
@@ -0,0 +1,28 @@
+# Function to print deprecation warnings, it will display a provided message as well as the calling function.
+# The deprecation message will only be displayed once per caller.
+# It is affected by the puppet setting 'strict', which can be set to :error, :off and :warning (default)
+#
+require "puppet_x/puppetlabs/stdlib/deprecation_state"
+
+Puppet::Functions.create_function(:deprecation) do
+  dispatch :deprecation do
+    param 'String', :message
+  end
+
+  def deprecation(message)
+    called_by = Kernel.caller.first
```

Comment:
There is a way to get calling puppet scope. No easy way to get file line as you need to find the calling instruction to do that. There is a ticket to improve so that calling file/line would be available to functions.
A 4.x function that wants access to calling scope must be an internal / system function. See the `defined` 4.x function in Puppet's code base for an example how to do an internal function, and how to get the calling scope parameter in the dispatcher.